### ISSUE TITLE
Instrumented compiler

### ISSUE THREAD

**Issue by Octachron** [Lereena/outreachy-work#3]
- FIrst, build the toy ppx with the in-tree compiler
- Second, add a target for the instrumented compiler, world.instrumented
- Third, run the testsuite with this instrumented compiler.

**Comment by shindere** [comment]
About task #1 whhat should work is the following scenario:

```shell
# Build the OCaml compiler first
./configure # configure the compiler
make world # Build the byte-code compiler
cd ppx # Go to your ppx directory
make # Build it with the freshly built compiler
```

**Comment by Lereena** [comment]
I changed the makefile for toy ppx and the main makefile for ocaml (this can be seen in my fork, but for convenience, I'll duplicate it here). 

Makefile for printexc_ppx:
```
ROOTROOT = ../..
ROOTDIR = ..

include $(ROOTDIR)/Makefile.config
include $(ROOTDIR)/Makefile.common

ocamlc := $(ROOTROOT)/runtime/ocamlrun $(ROOTROOT)/ocamlc

.PHONY: build
build:	
	cd printexc_ppx; $(ocamlc) -c printexc_ppx.ml
```

Changed part in the main Makefile:
```
.PHONY: world
world: coldstart
	$(MAKE) all
	cd ppx;	make build
```

Now when I try to build all by `$ make world` after `./configure`, I get the following error:
```
File "_none_", line 1:
Error: /usr/local/lib/ocaml/stdlib.cmi
       is not a compiled interface for this version of OCaml.
It seems to be for an older version of OCaml.
```

And I don't understand what the version is about, which is mentioned here

**Comment by Octachron** [comment]
I have commented directly on the commit creating the makefile. There is a few errors worth looking before this version mismatch error.

**Comment by shindere** [comment]
And I have opened a pull request with changes to perhaps give you some
inspiration :)

**Comment by Lereena** [comment]
If I'm not mistaken, tests are running by the ocamlopt, and now I need to build ocamlopt with ocamlc and ppx. It looks like `world.opt` is responsible for building it and `world.instrumentes` should be similar. I go to the`opt.opt` and see `ocamlc.opt` and `ocamlopt.opt`, but in comments specified that it is about compiling using ocamlopt. And I'm not sure in which part I need to make changes to get instrumented compiler.

**Comment by Octachron** [comment]
Most tests are run by both ocamlopt and ocamlc. In a first step, you can focus on building just an instrumented  ocamlc. One way to do it would be to add a PPX makefile variable, and then add this new flag to all invocation of ocamlc. In the standard mode, the variable should be empty. The instrumented target then should call the ordinary target while setting the PPX value to the ppx. (@shindere might have a better idea for this kind of configuration).

**Comment by Lereena** [comment]
Maybe I don’t understand the idea at all, but there must be something like that?
```
.PHONY: world.instrumented
world.instrumented: PPX = -ppx ./ppx/printexc_ppx
world.instrumented:
	$(MAKE) world
```

Then where in the ocamlc invocations do I need to put PPX? Is this generally related to the `world` taget?

**Comment by Octachron** [comment]
Well, you need to build the compiler with the ppx enabled. Did you try to find the rule that builds the cmo file from the ml files for instance?

**Comment by Lereena** [comment]
I'm not sure, but nothing but `.ml.cmo` looks so

**Comment by Octachron** [comment]
Yes, this is the right place to look. So you have an ocamlc invocation with two makefile variables. You should try to experiment a bit on where you should inject the PPX flag.

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/07 04:36 -0700):
> If I'm not mistaken, tests are running by the ocamlopt,

Well as Florian has said, most of the tests compiled with the two
compilers. In addition, if tests are compiled by only one compiler, most
of the time its ocamlc which is used because one is sure it's always
available, while ocamlopt may be disabled.

**Comment by shindere** [comment]
Florian Angeletti (2019/06/07 10:28 -0700):
> Most tests are run by both ocamlopt and ocamlc. In a first step, you
> can focus on building just an instrumented  ocamlc.

I agree

> One way to do it would be to add a PPX makefile variable, and then add
> this new flag to all invocation of ocamlc. In the standard mode, the
> variable should be empty. The instrumented target then should call the
> ordinary target while setting the PPX value to the ppx. (@shindere
> might have a better idea for this kind of configuration).

Well there is an OCAMLC variable which may be used for that, too, but
having a dedicated PPX variable seems right too.

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/07 11:39 -0700):
> Maybe I don’t understand the idea at all, but there must be something like that?
> ```
> .PHONY: world.instrumented
> world.instrumented: PPX = -ppx ./ppx/printexc_ppx
> world.instrumented:
> 	$(MAKE) world
> ```

Yeah in principle what you write makes sense. In practice, however, I
think the variable would not be inherited, the way you did it. It would
certainly work better if you do something like

```
.PHONY: instrumented-ocamlc
instrumented-ocamlc
        $(MAKE) PPX="-ppx ./ppx/printexc_ppx" ocamlc
```

**Comment by shindere** [comment]
Oxana Kostikova (2019/06/07 13:23 -0700):
> I'm not sure, but nothing but `.ml.cmo` looks so

Be aware that you may have to modify makefiles in the subdirectories,
too, not only the main makefile. You need to doble check this according
to what is needed to build the compiler (ocamlc) itself.

**Comment by Lereena** [comment]
It seems that form like this
```
.PHONY: instrumented-ocamlc
instrumented-ocamlc:
        $(MAKE) PPX="-ppx ./ppx/printexc_ppx" ocamlc
```
doesn't work at all. It produces an error of wrong `make` argument.
And if I'm not mistaken, following (and using PPX variable further) works as expected:
```
.PHONY: instrumented.ocamlc
instrumented.ocamlc: PPX=-ppx ./ppx/printexc_ppx
instrumented.ocamlc:
	$(MAKE) ocamlc
```
But the full picture of what is happening still does not add up.


I learned that `.SUFFIXES` produce implicit rules for creating files of source format to target. 
For my situation it means that .cmo files produced from .ml files using 
`$(CAMLC) $(COMPFLAGS) -c $<`
`CAMLC=$(BOOT_OCAMLC) -g -nostdlib -I boot -use-prims runtime/primitives`
`BOOT_OCAMLC = $(CAMLRUN) $(ROOTDIR)/boot/ocamlc`

And ocamlc is created by the
```
ocamlc: compilerlibs/ocamlcommon.cma compilerlibs/ocamlbytecomp.cma $(BYTESTART)
	$(CAMLC) $(LINKFLAGS) -compat-32 -o $@ $^
```
So I guess that I need to place PPX somewhere among these commands.

When I'm trying this:
```
.ml.cmo:
	$(CAMLC) $(COMPFLAGS) $(PPX) -c $<
```
or that:
`CAMLC=$(BOOT_OCAMLC) $(PPX) -g -nostdlib -I boot -use-prims runtime/primitives`

Then call `make world` and `make instrumented.ocamlc`, I'm getting
```
make ocamlc
make[1]: entering the directory «/home/lereena/ocaml»
make -C utils config.ml
make[2]: entering the directory «/home/lereena/ocaml/utils»
make[2]: «config.ml» does not require updating.
make[2]: entering the directory «/home/lereena/ocaml/utils»
make[1]: entering the directory «/home/lereena/ocaml»
```

It seems that it is necessary to inform the project somewhere that it is necessary to recompile ocamlc, 
but it is not very clear where, and whether this is a problem at all.

**Comment by Octachron** [comment]
Indeed, the root issue here is that the makefile is not aware that the cmo files have a dependency on the compilation flags. There are two options: we could either use a witness for the compilation under a ppx, or we could delete the pre-existing cmo files.
For this test, I propose to follow first the latter (simpler and hackish) option.
We can even be selective here, and delete just `typing/typetexp.cmo`.
To check that the right instrumentation is generated, you can test the produced compiler on
the following code
```OCaml
let f: 'a. (int as 'a) -> unit = ();;
```
Before I forget, to invoke the in-tree compiler by hand before installation, you can use:
```OCaml
./runtime/ocamlrun ./ocamlc -nostdlib -I stdlib ...
```

**Comment by Octachron** [comment]
Also, you should try to commit and push to your working branch as soon as you have made some progress.

**Comment by Lereena** [comment]
I've pushed what I have now. Not sure what should have happened on the test case, but an error occurred: 
```
Error: The universal type variable 'a cannot be generalized: it is bound to
       int.
```

**Comment by Octachron** [comment]
Is the current branch still `toy_ppx` (and not toy_ppx_fixed)?

**Comment by Lereena** [comment]
`toy_ppx` is current

**Comment by Octachron** [comment]
Concerning the test, what you should obtain is
```
raised Error on line 667
File "test.ml", line 1, characters 7-30:
1 | let f: 'a. (int as 'a) -> unit = fun _ -> ()
           ^^^^^^^^^^^^^^^^^^^^^^^
Error: The universal type variable 'a cannot be generalized: it is bound to
       int.
```
where the first line is displayed by the ppx.
You should first check if the makefile uses the ppx variable by reading the output of `make instrumented.ocamlc`.

**Comment by Lereena** [comment]
I've found where should I put PPX variable to make it update (and pushed).
I understand that my issue is that I am trying to apply ppx built for the older ast version to a more recent.
But I can't understand very much how is it possible to fix. I've just build a compiler to build a new version of the compiler.
I've build ppx. And now when I try to apply ppx, it says that compiler that built this ppx has an older version that is needed.
But where can I get more recent version if I just have a source? Rebuild ocamlc with just built ocamlc, then build ppx? Does it make any sense?

**Comment by Octachron** [comment]
Hm, I cannot reproduce this specific error. You may have older build artefact in your tree. You should delete the ppx executable and rebuild it. Also, after pulling the last version of the branch, doing `make world.instrumented` does not use the non-empty ppx flag. You should have a second look at the alternative proposed by @shindere earlier.

**Comment by Lereena** [comment]
Something super strange is happening, because my last version of using PPX variable worked, then stopped, then it worked again, and I definitely did not change anything except `instrumented.ocamlc` and rebuilding the `world`. Now I tried to remove the repository, download it again and change only the part of the instrumented.ocamlc to the version proposed by Sébastien, and again I get
```
Fatal error: the file './ppx/printexc_ppx' has not the right magic number: expected Caml1999X024, got Caml1999X025
```

**Comment by Octachron** [comment]
I think that I know how you get this error:  the ppx execution is picking up the wrong OCaml runtime. The easiest way to avoid this issue is to pack the (right) runtime with the ppx bytecode executable by adding `-custom` to the ppx's Makefile.

**Comment by Lereena** [comment]
It seems to work. But, for some reason, each raise has two printed lines. I ran tests with this compiler, got about twice as many errors as I should have :D

**Comment by Octachron** [comment]
Great! If you want, you could try to see if you can catch an error that appears in the `typing/typetexp.ml` file but is not detected when you run the testsuite.

**Comment by Lereena** [comment]
I tried to see what errors arise in `typing/typetexp.ml`, but, to be honest, everything in it looks rather confusing and I have no idea how to debug it. I do not even really understand what is happening in the tests that check it.

I also thought about what I need to use `bisect_ppx` and see how it is used in some projects. How problematic is the use of dune when building compilers? Can we use instructions from Makefile that is attached by default to `bisect_ppx`?

**Comment by Octachron** [comment]
I was more thinking at looking at the displayed exception, checking that they correspond to a raised exception in the code source, and then look for an exception that does not appear in the displayed exceptions when running the testsuite. It might be easier if you edited the ppx to write to a fixed log file rather than stderr.

We cannot really use `dune` in the compiler, but writing a short Makefile is not a big obstacle. Me or @shindere might take the time to write such a Makefile if necessary.

**Comment by Lereena** [comment]
I was stuck in the process of knowing what to do. In theory, tests check for errors in the definition of common types and polymorphic variants, and a kind of type table is sorted out at the place of occurrence of an error and an error is constructed if it does not contain the expected types. But how to find out what happens in principle when this code is working is not very clear. Is it possible to get a stack trace just for this fragment or something like to find out the direction in which to look? Or maybe I'm trying to do not that kind of thing that is needed?

**Comment by Octachron** [comment]
Ok, let's slow down a bit. I am not sure what you are trying to do. What I was proposing was:

* modify your toy_ppx to print to a log file rather than stdout
* run the testsuite
* look at the log file generated by the ppx
* each line in the log file should correspond a line in `typing/typetexp.ml` where an exception is raised.
* then, you can try to check the reverse: does each line where an exception is raised appear in your log file? (The answer is no).

**Comment by Lereena** [comment]
I do not really understand what is the point of modifying ppx for output to the log, because if I redirect the output to a file, then, in theory, I get the same thing? I am getting this output:
```
...
 ... testing 'typetexp_errors.ml' with 1 (expect) => raised Error on line 800
raised Error on line 800
raised Error on line 631
raised Error on line 631
raised Error on line 800
raised Error on line 800
raised Error on line 631
raised Error on line 631
passed
...
```

**Comment by Octachron** [comment]
It was to avoid mixing ocamltest output with the ppx output, but it is not necessary indeed.

**Comment by Lereena** [comment]
In that case, I need to run the functions from `typing/typetexp.ml` separately from the tests and see where exceptions occur, right?

**Comment by Octachron** [comment]
Nope, you just need to read the functions from `typing/typetexp.ml`. Or even simpler: `cat typing/typetexp.ml | grep -n raise`

**Comment by Lereena** [comment]
Hmm, for some reason I made a not very clever assumption that the point is that only part of the exceptions are called, so I have to figure out which ones, and I have been focused on it all this time. Now I understand that the point is that they are not explored in tests. Thank you very much.

**Comment by Lereena** [comment]
In this case, I’m expected to somehow add tests that will check something from the uncovered part?

**Comment by Octachron** [comment]
At this stage, not really except if you really want to. The idea was more to have a practical and small implementation of a ppx that can detect some missing test before switching to `bisect_ppx`.
